openstack_hypervisor/fileserver/server.py

```diff
@@ -0,0 +1,434 @@
+# SPDX-FileCopyrightText: 2025 - Canonical Ltd
+# SPDX-License-Identifier: Apache-2.0
+"""WSGI server for remote file transfer operations (threaded backend).
+
+This module exposes a minimal WebOb-based WSGI application, typically
+hosted with ``oslo_service.wsgi.Server``. It implements endpoints for
+basic filesystem operations, chunked uploads (with optional gzip
+compression), and file downloads. TLS support is configured via
+``oslo_service.sslutils`` and ``oslo_config``.
+"""
+
+import errno
+import gzip as _gzip
+import json
+import os
+import ssl
+import threading
+import uuid
+from pathlib import Path
+from socketserver import ThreadingMixIn
+from typing import Any, Dict
+from wsgiref.simple_server import WSGIServer, make_server
+
+from oslo_config import cfg
+from oslo_log import log as logging
+from oslo_service import service, sslutils
+from webob import Request, Response
+
+from .utils import (
+    CHUNK_READ_SIZE,
+    BadRequestError,
+    assemble_chunks_to_file,
+    cleanup_upload,
+    compute_sha256,
+    get_snap_common,
+    read_meta_json,
+    uploads_root,
+    write_meta_json,
+)
+
+LOG = logging.getLogger(__name__)
+
+
+fileserver_opts = [
+    cfg.StrOpt(
+        "host",
+        default=os.environ.get("FILESERVER_HOST", "0.0.0.0"),
+        help="Listen address for the file server",
+    ),
+    cfg.IntOpt(
+        "port",
+        default=int(os.environ.get("FILESERVER_PORT", "8099")),
+        min=1,
+        help="TCP listen port for the file server",
+    ),
+    cfg.StrOpt(
+        "sandbox_root",
+        default=os.environ.get("SNAP_COMMON"),
+        help="Base directory for sandboxed filesystem operations",
+    ),
+]
+
+CONF = cfg.CONF
+CONF.register_opts(fileserver_opts, group="fileserver")
+sslutils.register_opts(CONF)
+
+
+def _resolve_sandboxed_path(path_str: str) -> Path | None:
+    """Resolve a user-supplied path within the sandbox (SNAP_COMMON).
+
+    Returns an absolute Path when the path is within SNAP_COMMON. If
+    SNAP_COMMON is not set, returns the raw path as-is (for tests/dev).
+    Returns None if the resolved path escapes the sandbox.
+    """
+    base_str = os.environ.get("SNAP_COMMON") or CONF.fileserver.sandbox_root
```

Comment:
SNAP_COMMON is the default as defined a few lines earlier, this feels redundant.